src/fairseq2/utils/parquet_dataloader.py

```diff
@@ -0,0 +1,518 @@
+    def __init__(
+        self,
+        config: ParquetBasicDataloaderConfig,
+    ) -> None:
+        self.config = config
+
+        # split_row_groups=True is not supported yet
+        self.source_ds = pq.ParquetDataset(
+            self.config.parquet_path, validate_schema=True, filters=self.config.filters
+        )
+
+        self.columns = self.config.columns or self.source_ds.schema.names
+        assert set(self.columns).issubset(set(self.source_ds.schema.names))
+
+        if self.config.order_by is not None:
+            assert self.config.order_by in self.source_ds.schema.names
+
+        partitioning_keys = (
+            [
+                name
+                for (name, dd) in zip(
+                    self.source_ds.partitioning.schema.names,
+                    self.source_ds.partitioning.dictionaries,
+                )
+                if dd is not None
+            ]
+            if self.source_ds.partitioning
+            else []
+        )
+        columns_wo_partition_keys = [
+            col for col in self.columns if col not in partitioning_keys
+        ]
+
+        if self.config.order_by is not None:
+            self._columns_to_read = sorted(
+                set(columns_wo_partition_keys) | set([self.config.order_by])
+            )
+        else:
+            self._columns_to_read = columns_wo_partition_keys
+
+        if self.config.split_to_row_groups:
+            self._all_fragments = [
+                piece
+                for fragment in self.source_ds._dataset.get_fragments(
+                    self.config.filters
+                )
+                for piece in fragment.split_by_row_group()
+            ]
+        else:
+            self._all_fragments = list(
+                self.source_ds._dataset.get_fragments(self.config.filters)
+            )
+
+    # def __repr__(self) -> str:
+    #     """ count rows and other simple stats """"
+
+    # def head(self, nb:int=5) -> pa.Table:
+    #     return self.source_ds.head(nb=5)
+
+    def schema(self) -> pa.Schema:
+        """
+        Returns the full schema of the parquet datasource
+        """
+        return self.source_ds.schema
+
+    @staticmethod
+    def _add_partitioning_values(
+        table: pa.Table, fragment: pa.dataset.Fragment
+    ) -> pa.Table:
+        for key, val in get_partition_keys(fragment.partition_expression).items():
+            values = pd.Series([val] * len(table), dtype="category")
+            table = table.append_column(key, pa.Array.from_pandas(values))
+        return table
+
+    @staticmethod
+    def _sanitize_dict(table: pa.Table) -> pa.Table:
+        for i, (name, type) in enumerate(zip(table.schema.names, table.schema.types)):
+            if pa.types.is_dictionary(type):
+                ca = table[name].cast(type.value_type)
+                table = table.remove_column(i)
+                table = table.append_column(name, ca)
+        return table
+
+    @staticmethod
+    def compute_length_splits(
```

Comment:
what's the benefice of having this logic here and not in the pipeline like we do with the other fairseq2 dataloader ?

Reply:
yes, indeed a good point ! can you please show me how to realize this logic with fairseq2 primitives ?